prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+	OFI_IBV_MTU_TYPE(256, bytes);
+	OFI_IBV_MTU_TYPE(512, bytes);
+	OFI_IBV_MTU_TYPE(1024, bytes);
+	OFI_IBV_MTU_TYPE(2048, bytes);
+	OFI_IBV_MTU_TYPE(4096, bytes);
+	}
+
+	return bytes;
+}
+#define OFI_VERBS_UD_MTU_TYPE_TO_LEN(mtu_type)	\
+do {						\
+	switch ((enum ibv_mtu) mtu_type) {	\
+	case IBV_MTU_256:			\
+	case IBV_MTU_512:			\
+	case IBV_MTU_1024:			\
+	case IBV_MTU_2048:			\
+	case IBV_MTU_4096:			\
+	default:				\
+		assert(0);			\
+	}					\
+} while (0)
+
+struct ofi_verbs_ud_buf_pool;
+
+typedef void(*ofi_verbs_ud_pool_entry_cancel_hndlr) (struct ofi_verbs_ud_buf_pool *);
+
+struct ofi_verbs_ud_buf_pool {
+	struct util_buf_pool	*pool;
+	struct dlist_entry	buf_list;
+
+	ofi_verbs_ud_pool_entry_cancel_hndlr cancel_hndlr;
+};
+
+/*
+ * fields of Infiniband packet headers that are used to
+ * represent OFI EP address
+ * - LRH (Local Route Header) - Link Layer:
+ *   - LID - destination Local Identifier
+ *   - SL - Service Level
+ * - GRH (Global Route Header) - Network Layer:
+ *   - GID - destination Global Identifier
+ * - BTH (Base Transport Header) - Transport Layer:
+ *   - QPN - destination Queue Oair number
+ *   - PSN - Packet Serial Number
+ *   - P_key - Partition Key
+ *
+ * Note: DON'T change the placement of the fields in the structure.
+ *       The placement is to keep structure size = 256 bits (32 byte).
+ */
+struct ofi_ib_ud_ep_name {
+	union ibv_gid	gid; /* 64-bit GUID + 64-bit EUI - GRH */
+
+	uint32_t	qpn; /* BTH */
+	uint32_t	psn; /* SQ PSN - BTH */
+
+	uint16_t	lid; /* LRH */
+	uint16_t	pkey; /* BTH */
+	uint16_t	service; /* for NS src addr, 0 means any */
+
+	uint8_t 	sl; /* LRH */
+	uint8_t		padding; /* forced padding to 256 bits (32 byte) */
+}; /* 256 bits */
+
+struct ofi_verbs_ud_fabric {
+	struct util_fabric	util_fabric;
+	struct util_ns		name_server;
+};
+
+struct ofi_verbs_ud_domain {
+	struct util_domain	util_domain;
+	struct util_av		*av;
+    	struct fi_info		*info;
+	struct ibv_comp_channel	*ibv_comp_channel;
+	struct ibv_context	*ibv_device_ctx;
+	int			port_number;
+	struct ibv_port_attr	*port_attr;
+	struct ibv_pd		*ibv_pd;
+	/* NULL if gl_data::separate_send_recv_cq = 1 */
+	struct ibv_cq		*ibv_shared_cq;
+};
+
+struct ofi_verbs_ud_ep {
+	struct util_ep			util_ep;
+	struct ibv_qp			*ibv_qp;
+	struct fi_info			*info;
+	struct ofi_verbs_ud_av		*av;
+	struct ofi_verbs_ud_domain	*domain;
+	struct ibv_cq			*ibv_shared_cq;
+	struct ibv_cq			*ibv_send_cq;
+    	struct ibv_cq			*ibv_recv_cq;
+	struct ofi_verbs_ud_buf_pool	grh_pool;
+	struct ofi_ib_ud_ep_name	ep_name;
+	int				service;
+	int				ep_flags;
+	struct ibv_wc			*multiple_wcs;
+};
+
+struct ofi_verbs_ud_cq {
+	struct util_cq	util_cq;
+	struct ibv_cq	*ibv_cq;
+};
+
+struct ofi_verbs_ud_mem_desc {
+	struct fid_mr			mr_fid;
+	struct ibv_mr			*mr;
+	struct ofi_verbs_ud_domain	*domain;
+};
+
+struct ofi_verbs_ud_av {
+	struct util_av	util_av;
+	RbtHandle	addr_map;
+};
+
+struct ofi_verbs_ud_av_entry {
+	struct ofi_ib_ud_ep_name	*addr;
```

Comment:
Is it better to store an 8-byte pointer to a 32-byte address, or just embed the address?  It looks like the cross-over point for storage is 75% capacity.  Below that a pointer is better, above that the address is better.